lib/appium.js

```diff
@@ -178,24 +183,30 @@ class AppiumDriver extends BaseDriver {
       }
     }
 
-    let runningDriversData;
-    try {
-      runningDriversData = this.curSessionDataForDriver(InnerDriver);
-    } catch (e) {
-      throw new errors.SessionNotCreatedError(e.message);
-    }
-
-    let innerSessionId, dCaps;
+    let runningDriversData, otherPendingDriversData;
     let d = new InnerDriver(this.args);
-    try {
+    await sessionCreationLock.acquire(AppiumDriver.name, async () => {
+      try {
+        runningDriversData = this.curSessionDataForDriver(InnerDriver);
+      } catch (e) {
+        throw new errors.SessionNotCreatedError(e.message);
+      }
+
       this.pendingDrivers[InnerDriver.name] = this.pendingDrivers[InnerDriver.name] || [];
-      const otherPendingDriversData = this.pendingDrivers[InnerDriver.name].map((drv) => drv.driverData);
+      otherPendingDriversData = this.pendingDrivers[InnerDriver.name].map((drv) => drv.driverData);
       this.pendingDrivers[InnerDriver.name].push(d);
+    });
+    let innerSessionId, dCaps;
+    try {
       [innerSessionId, dCaps] = await d.createSession(caps, reqCaps, [...runningDriversData, ...otherPendingDriversData]);
     } finally {
-      _.pull(this.pendingDrivers[InnerDriver.name], d);
+      await sessionCreationLock.acquire(AppiumDriver.name, async () => {
```

Comment:
same for this. or is it that we want to lock both sides of the `await d.createSession` call?